File: 백준/레벨Platinum V/2150. Strongly Connected Component/Strongly Connected Component.cc
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> graph;
// SCC 후보 스택, 정점의 스택 포함 여부
stack<int> sccStack;
vector<bool> inStack;
// DFS 방문순서, 카운터
vector<int> visited;
int visitedCount;
// SCC 목록
vector<vector<int>> sccList;

int dfs(int u) {
    // low : u에서 역방향으로 거슬러 갈 수 있는 가장 작은 방문순서
    visited[u] = visitedCount;
    int low = visitedCount++;

    // 스택에 u를 추가
    sccStack.push(u);
    inStack[u] = true;

    // 인접 정점을 탐색
    for (int v : graph[u]) {
        // v 방문 전이면 dfs(v)후 low 갱신
        if (visited[v] == -1) {
            low = min(low, dfs(v)); // 두번째 low : 자식 정점의 방문 순서
        }
        // 방문했지만 SCC 후보라면 low 갱신
        else if (inStack[v]) {
            low = min(low, visited[v]); // 세번째 low : 스택에 있는 방문 순서
        }
    }

    // low가 방문순서와 동일 : 먼저 방문된 노드로 돌아갈 수 없어서 여기가 SCC 루트
    if (low == visited[u]) {
        vector<int> curSCC;

        // SCC 구성하기
        while (true) {
            int top = sccStack.top();
            sccStack.pop();
            inStack[top] = false;
            curSCC.push_back(top);
            if (top == u) break;
        }
        sort(curSCC.begin(), curSCC.end());
        sccList.push_back(curSCC);
    }

    return low;
}

void solution() {
    // 정방향, 역방향 인접리스트 초기화
    int V, E;
    cin >> V >> E;
    graph.assign(V + 1, vector<int>());
    
    for (int i = 0; i < E; i++) {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
    }

    sccStack = stack<int>();
    inStack.assign(V + 1, false);
    visited.assign(V + 1, -1);
    visitedCount = 0;
    sccList.clear();

    // 모든 정점에 대해 DFS
    for (int u = 1; u <= V; u++) {
        if (visited[u] == -1) dfs(u);
    }

    // 첫줄에 SCC의 개수 K를 출력
    // 다음 K개의 줄에는 각 줄에 하나의 SCC에 속한 정점의 번호를 출력
    sort(sccList.begin(), sccList.end(), [](const vector<int>& a, const vector<int>& b) {
        return a[0] < b[0];
    });

    cout << sccList.size() << '\n';
    for (const auto& scc : sccList) {
        for (int v : scc) {
            cout << v << ' ';
        }
        cout << "-1\n";
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    solution();
}

File: 백준/레벨Platinum V/2150. Strongly Connected Component/Strongly Connected Component.java
import java.io.*;
import java.util.*;

class Kosaraju {
    static int V, E;
    static List<List<Integer>> fGraph;
    static List<List<Integer>> rGraph;
    static boolean[] visited;
    static Deque<Integer> orderStack;

    public void solution() throws IOException {
        // 정방향, 역방향 인접리스트 초기화
        V = nextInt();
        E = nextInt();
        fGraph = new ArrayList<>();
        rGraph = new ArrayList<>();

        for (int i = 0; i <= V; i++) {
            fGraph.add(new ArrayList<>());
            rGraph.add(new ArrayList<>());
        }

        for (int i = 0; i < E; i++) {
            int u = nextInt(), v = nextInt();
            fGraph.get(u).add(v);
            rGraph.get(v).add(u);
        }

        // 모든 정점에 대해 정방향 DFS
        visited = new boolean[V + 1];
        orderStack = new LinkedList<>();

        for (int u = 1; u <= V; u++) {
            if (!visited[u]) fDFS(u);
        }

        // 방문 정점에 대해 역방향 DFS
        visited = new boolean[V + 1];
        List<List<Integer>> sccList = new ArrayList<>();

        while (!orderStack.isEmpty()) {
            int v = orderStack.pop();
            if (!visited[v]) {
                List<Integer> curSCC = new LinkedList<>();
                rDFS(v, curSCC);
                Collections.sort(curSCC);
                sccList.add(curSCC);
            }
          }
        
        // 첫줄에 SCC의 개수 K를 출력
        // 다음 K개의 줄에는 각 줄에 하나의 SCC에 속한 정점의 번호를 출력
        Collections.sort(sccList, (a, b) -> a.get(0) - b.get(0));

        StringBuilder sb = new StringBuilder();
        sb.append(sccList.size()).append('\n');
        for (List<Integer> scc : sccList) {
            for (int v : scc) {
                sb.append(v).append(' ');
            }
            sb.append("-1\n");
        }        

        System.out.println(sb);
    }

    static void fDFS(int u) {
        visited[u] = true;
        for (int v : fGraph.get(u)) {
            if (!visited[v]) fDFS(v);
        }
        // 정점방문 순서를 추가
        orderStack.push(u);        
    }

    static void rDFS(int v, List<Integer> curSCC) {
        visited[v] = true;
        // 현재 SCC를 구성
        curSCC.add(v);
        for (int u : rGraph.get(v)) {
          if (!visited[u]) rDFS(u, curSCC);
        }       
    }

    private static int nextInt() throws IOException {
        int n = 0;
        int c;
        while ((c = System.in.read()) <= 32);
        do {
            n = n * 10 + (c - '0');
        } while ((c = System.in.read()) > 32);
        return n;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        new Kosaraju().solution();
    }
}
